File: recap-subworker/recap_subworker/infra/otel.py
"""
ABOUTME: OpenTelemetry provider for recap-subworker service.
ABOUTME: Handles trace and log export to OTLP endpoint with ADR 98 compliance.
"""

import os
from collections.abc import Callable

from opentelemetry import trace
from opentelemetry._logs import set_logger_provider
from opentelemetry.exporter.otlp.proto.http._log_exporter import OTLPLogExporter
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk._logs import LoggerProvider
from opentelemetry.sdk._logs.export import BatchLogRecordProcessor
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor


class OTelConfig:
    """OpenTelemetry configuration from environment variables."""

    def __init__(self):
        self.service_name = os.getenv("OTEL_SERVICE_NAME", "recap-subworker")
        self.service_version = os.getenv("SERVICE_VERSION", "0.1.0")
        self.environment = os.getenv("DEPLOYMENT_ENV", "development")
        self.otlp_endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://localhost:4318")
        self.enabled = os.getenv("OTEL_ENABLED", "true").lower() == "true"


def init_otel_provider(config: OTelConfig | None = None) -> Callable[[], None]:
    """
    Initialize OpenTelemetry providers for tracing and logging.

    Args:
        config: Optional OTel configuration. If None, reads from environment.

    Returns:
        A shutdown function to clean up providers.
    """
    if config is None:
        config = OTelConfig()

    if not config.enabled:
        return lambda: None

    # Create resource with service information
    resource = Resource.create(
        {
            "service.name": config.service_name,
            "service.version": config.service_version,
            "deployment.environment": config.environment,
        }
    )

    # Initialize Tracer Provider
    tracer_provider = TracerProvider(resource=resource)
    trace_exporter = OTLPSpanExporter(
        endpoint=f"{config.otlp_endpoint}/v1/traces",
    )
    tracer_provider.add_span_processor(BatchSpanProcessor(trace_exporter))
    trace.set_tracer_provider(tracer_provider)

    # Initialize Logger Provider
    logger_provider = LoggerProvider(resource=resource)
    log_exporter = OTLPLogExporter(
        endpoint=f"{config.otlp_endpoint}/v1/logs",
    )
    logger_provider.add_log_record_processor(BatchLogRecordProcessor(log_exporter))
    set_logger_provider(logger_provider)

    def shutdown():
        """Shutdown OTel providers gracefully."""
        tracer_provider.shutdown()
        logger_provider.shutdown()

    return shutdown

File: alt-perf/src/utils/otel.ts
/**
 * OpenTelemetry provider for alt-perf CLI tool.
 * Handles optional OTLP log export for observability.
 */

import { logs, SeverityNumber } from "@opentelemetry/api-logs";
import { LoggerProvider, BatchLogRecordProcessor } from "@opentelemetry/sdk-logs";
import { OTLPLogExporter } from "@opentelemetry/exporter-logs-otlp-http";
import { Resource } from "@opentelemetry/resources";
import {
  SEMRESATTRS_SERVICE_NAME,
  SEMRESATTRS_SERVICE_VERSION,
  SEMRESATTRS_DEPLOYMENT_ENVIRONMENT,
} from "@opentelemetry/semantic-conventions";

/**
 * OTel configuration from environment variables
 */
export interface OTelConfig {
  serviceName: string;
  serviceVersion: string;
  environment: string;
  otlpEndpoint: string;
  enabled: boolean;
}

/**
 * Get OTel configuration from environment
 */
export function getOTelConfig(): OTelConfig {
  return {
    serviceName: Deno.env.get("OTEL_SERVICE_NAME") || "alt-perf",
    serviceVersion: Deno.env.get("SERVICE_VERSION") || "1.0.0",
    environment: Deno.env.get("DEPLOYMENT_ENV") || "development",
    otlpEndpoint: Deno.env.get("OTEL_EXPORTER_OTLP_ENDPOINT") || "http://localhost:4318",
    // Default to false for CLI tool (opt-in)
    enabled: (Deno.env.get("OTEL_ENABLED") || "false").toLowerCase() === "true",
  };
}

let loggerProvider: LoggerProvider | null = null;
let otelLogger: ReturnType<typeof logs.getLogger> | null = null;

/**
 * Initialize OpenTelemetry logger provider
 */
export function initOTelProvider(config?: OTelConfig): () => void {
  const cfg = config || getOTelConfig();

  if (!cfg.enabled) {
    return () => {};
  }

  // Create resource with service information
  const resource = new Resource({
    [SEMRESATTRS_SERVICE_NAME]: cfg.serviceName,
    [SEMRESATTRS_SERVICE_VERSION]: cfg.serviceVersion,
    [SEMRESATTRS_DEPLOYMENT_ENVIRONMENT]: cfg.environment,
  });

  // Create OTLP log exporter
  const logExporter = new OTLPLogExporter({
    url: `${cfg.otlpEndpoint}/v1/logs`,
  });

  // Create logger provider
  loggerProvider = new LoggerProvider({
    resource,
  });

  loggerProvider.addLogRecordProcessor(new BatchLogRecordProcessor(logExporter));

  // Set as global provider
  logs.setGlobalLoggerProvider(loggerProvider);

  // Get logger instance
  otelLogger = logs.getLogger("alt-perf");

  return () => {
    if (loggerProvider) {
      loggerProvider.shutdown();
      loggerProvider = null;
      otelLogger = null;
    }
  };
}

/**
 * Map log level string to OTel severity number
 */
function levelToSeverity(level: string): SeverityNumber {
  switch (level.toLowerCase()) {
    case "error":
      return SeverityNumber.ERROR;
    case "warn":
    case "warning":
      return SeverityNumber.WARN;
    case "info":
      return SeverityNumber.INFO;
    case "debug":
      return SeverityNumber.DEBUG;
    default:
      return SeverityNumber.INFO;
  }
}

/**
 * Emit a log record to OTel
 */
export function emitOTelLog(
  level: string,
  message: string,
  attributes: Record<string, string | number | boolean> = {}
): void {
  if (!otelLogger) {
    return;
  }

  // Filter out undefined/null values and ensure valid types
  const cleanAttributes: Record<string, string | number | boolean> = {};
  for (const [key, value] of Object.entries(attributes)) {
    if (value !== undefined && value !== null) {
      if (typeof value === "string" || typeof value === "number" || typeof value === "boolean") {
        cleanAttributes[key] = value;
      } else {
        cleanAttributes[key] = String(value);
      }
    }
  }

  otelLogger.emit({
    severityNumber: levelToSeverity(level),
    severityText: level.toUpperCase(),
    body: message,
    attributes: cleanAttributes,
  });
}

/**
 * Check if OTel is enabled and initialized
 */
export function isOTelEnabled(): boolean {
  return otelLogger !== null;
}
